File: js/euler62.js
function euler62(){
    // a cube must have at least 3 digits to have 5 or more permutations
    var maxLen = 3;
    var minCube = Number.MAX_VALUE;
    while (minCube == Number.MAX_VALUE){
        // group cubes by sorted digit strings
        let cubeDigits = new Map();
        for (let n = Math.ceil(Math.cbrt(10 ** (maxLen - 1))); n <= Math.floor(Math.cbrt(10 ** maxLen)); n++){
            let cube = n ** 3;
            // get sorted digit string and store as key in map
            let digits = cube.toString().split('').sort().join('');
            if (!cubeDigits.has(digits)){
                cubeDigits.set(digits, new Set());
            }
            cubeDigits.get(digits).add(cube);
        }
        for (let cubes of cubeDigits.values()){
            if (cubes.size == 5){
                // find minimum cube in cubes and update minCube
                minCube = Math.min(minCube, ...cubes);
            }
        }
        maxLen++;
    }
    return minCube;
}

console.log(euler62());

File: go/euler62.go
package main
import (
    "fmt"
    "math"
    "sort"
    "strings"
    "strconv"
)

func sortDigits(n int) string {
    s := strings.Split(strconv.Itoa(n), "")
    sort.Strings(s)
    digits := strings.Join(s, "")
    return digits
}

func main() {
    res := math.MaxInt64
    // a number has at least 5 permutations if it has at least 3 digits
    maxLen := 3
    for res == math.MaxInt64 {
        cubeDigits := make(map[string][]int)
        start := math.Ceil(math.Cbrt(math.Pow(10, float64(maxLen - 1))))
        end := math.Floor(math.Cbrt(math.Pow(10, float64(maxLen))))
        for n := int(start); n <= int(end); n++ {
            cube := n * n * n
            digits := sortDigits(cube)
            cubeDigits[digits] = append(cubeDigits[digits], cube)
        }
        for _, group := range cubeDigits {
            if len(group) == 5 && group[0] < res {
                res = group[0]
            }
        }
        maxLen++
    }
    fmt.Println(res)
}
